File: java/lib/src/main/java/com/svix/models/RedshiftConfigIn.java
// This file is @generated
package com.svix.models;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.svix.Utils;

import lombok.EqualsAndHashCode;
import lombok.ToString;

@ToString
@EqualsAndHashCode
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonAutoDetect(getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class RedshiftConfigIn {
    @JsonProperty private String accessKeyId;
    @JsonProperty private String secretAccessKey;
    @JsonProperty private String region;
    @JsonProperty private String clusterIdentifier;
    @JsonProperty private String dbUser;
    @JsonProperty private String workgroupName;
    @JsonProperty private String dbName;
    @JsonProperty private String schemaName;
    @JsonProperty private String tableName;

    public RedshiftConfigIn() {}

    public RedshiftConfigIn accessKeyId(String accessKeyId) {
        this.accessKeyId = accessKeyId;
        return this;
    }

    /**
     * Access key ID.
     *
     * <p>Currently a required field, but marked as optional because we may add different
     * authentication in the future.
     *
     * @return accessKeyId
     */
    @javax.annotation.Nullable
    public String getAccessKeyId() {
        return accessKeyId;
    }

    public void setAccessKeyId(String accessKeyId) {
        this.accessKeyId = accessKeyId;
    }

    public RedshiftConfigIn secretAccessKey(String secretAccessKey) {
        this.secretAccessKey = secretAccessKey;
        return this;
    }

    /**
     * Secret access key.
     *
     * <p>Currently a required field, but marked as optional because we may add different
     * authentication in the future.
     *
     * @return secretAccessKey
     */
    @javax.annotation.Nullable
    public String getSecretAccessKey() {
        return secretAccessKey;
    }

    public void setSecretAccessKey(String secretAccessKey) {
        this.secretAccessKey = secretAccessKey;
    }

    public RedshiftConfigIn region(String region) {
        this.region = region;
        return this;
    }

    /**
     * The region of the Redshift DB.
     *
     * <p>Currently a required field, but marked as optional because we may infer it from other
     * fields in the future.
     *
     * @return region
     */
    @javax.annotation.Nullable
    public String getRegion() {
        return region;
    }

    public void setRegion(String region) {
        this.region = region;
    }

    public RedshiftConfigIn clusterIdentifier(String clusterIdentifier) {
        this.clusterIdentifier = clusterIdentifier;
        return this;
    }

    /**
     * Required for provisioned clusters.
     *
     * @return clusterIdentifier
     */
    @javax.annotation.Nullable
    public String getClusterIdentifier() {
        return clusterIdentifier;
    }

    public void setClusterIdentifier(String clusterIdentifier) {
        this.clusterIdentifier = clusterIdentifier;
    }

    public RedshiftConfigIn dbUser(String dbUser) {
        this.dbUser = dbUser;
        return this;
    }

    /**
     * Required for provisioned clusters.
     *
     * @return dbUser
     */
    @javax.annotation.Nullable
    public String getDbUser() {
        return dbUser;
    }

    public void setDbUser(String dbUser) {
        this.dbUser = dbUser;
    }

    public RedshiftConfigIn workgroupName(String workgroupName) {
        this.workgroupName = workgroupName;
        return this;
    }

    /**
     * Required for Redshift Serverless.
     *
     * @return workgroupName
     */
    @javax.annotation.Nullable
    public String getWorkgroupName() {
        return workgroupName;
    }

    public void setWorkgroupName(String workgroupName) {
        this.workgroupName = workgroupName;
    }

    public RedshiftConfigIn dbName(String dbName) {
        this.dbName = dbName;
        return this;
    }

    /**
     * Database name.
     *
     * <p>Only required if not using transformations.
     *
     * @return dbName
     */
    @javax.annotation.Nullable
    public String getDbName() {
        return dbName;
    }

    public void setDbName(String dbName) {
        this.dbName = dbName;
    }

    public RedshiftConfigIn schemaName(String schemaName) {
        this.schemaName = schemaName;
        return this;
    }

    /**
     * Schema name.
     *
     * <p>Only used if not using transformations.
     *
     * @return schemaName
     */
    @javax.annotation.Nullable
    public String getSchemaName() {
        return schemaName;
    }

    public void setSchemaName(String schemaName) {
        this.schemaName = schemaName;
    }

    public RedshiftConfigIn tableName(String tableName) {
        this.tableName = tableName;
        return this;
    }

    /**
     * Table name.
     *
     * <p>Only required if not using transformations.
     *
     * @return tableName
     */
    @javax.annotation.Nullable
    public String getTableName() {
        return tableName;
    }

    public void setTableName(String tableName) {
        this.tableName = tableName;
    }

    /**
     * Create an instance of RedshiftConfigIn given an JSON string
     *
     * @param jsonString JSON string
     * @return An instance of RedshiftConfigIn
     * @throws JsonProcessingException if the JSON string is invalid with respect to
     *     RedshiftConfigIn
     */
    public static RedshiftConfigIn fromJson(String jsonString) throws JsonProcessingException {
        return Utils.getObjectMapper().readValue(jsonString, RedshiftConfigIn.class);
    }

    /**
     * Convert an instance of RedshiftConfigIn to an JSON string
     *
     * @return JSON string
     */
    public String toJson() throws JsonProcessingException {
        return Utils.getObjectMapper().writeValueAsString(this);
    }
}

File: php/src/Models/RedshiftConfigIn.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Models;

/**
 * Configuration parameters for defining a Redshift sink.
 *
 * For provisioned clusters, set `cluster_identifier` and `db_user`. For Redshift Serverless, set `workgroup_name`.
 */
class RedshiftConfigIn implements \JsonSerializable
{
    private array $setFields = [];

    /**
     * @param string|null $accessKeyId Access key ID.
     *
     * Currently a required field, but marked as optional because we may add different authentication in the future.
     * @param string|null $secretAccessKey Secret access key.
     *
     * Currently a required field, but marked as optional because we may add different authentication in the future.
     * @param string|null $region The region of the Redshift DB.
     *
     * Currently a required field, but marked as optional because we may infer it from other fields in the future.
     * @param string|null $clusterIdentifier required for provisioned clusters
     * @param string|null $dbUser            required for provisioned clusters
     * @param string|null $workgroupName     required for Redshift Serverless
     * @param string|null $dbName            Database name.
     *
     * Only required if not using transformations.
     * @param string|null $schemaName Schema name.
     *
     * Only used if not using transformations.
     * @param string|null $tableName Table name.
     *
     * Only required if not using transformations.
     */
    private function __construct(
        public readonly ?string $accessKeyId = null,
        public readonly ?string $secretAccessKey = null,
        public readonly ?string $region = null,
        public readonly ?string $clusterIdentifier = null,
        public readonly ?string $dbUser = null,
        public readonly ?string $workgroupName = null,
        public readonly ?string $dbName = null,
        public readonly ?string $schemaName = null,
        public readonly ?string $tableName = null,
        array $setFields = [],
    ) {
        $this->setFields = $setFields;
    }

    /**
     * Create an instance of RedshiftConfigIn with required fields.
     */
    public static function create(
    ): self {
        return new self(
            accessKeyId: null,
            secretAccessKey: null,
            region: null,
            clusterIdentifier: null,
            dbUser: null,
            workgroupName: null,
            dbName: null,
            schemaName: null,
            tableName: null,
            setFields: []
        );
    }

    public function withAccessKeyId(?string $accessKeyId): self
    {
        $setFields = $this->setFields;
        $setFields['accessKeyId'] = true;

        return new self(
            accessKeyId: $accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $this->region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $this->workgroupName,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withSecretAccessKey(?string $secretAccessKey): self
    {
        $setFields = $this->setFields;
        $setFields['secretAccessKey'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $secretAccessKey,
            region: $this->region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $this->workgroupName,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withRegion(?string $region): self
    {
        $setFields = $this->setFields;
        $setFields['region'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $this->workgroupName,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withClusterIdentifier(?string $clusterIdentifier): self
    {
        $setFields = $this->setFields;
        $setFields['clusterIdentifier'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $this->region,
            clusterIdentifier: $clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $this->workgroupName,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withDbUser(?string $dbUser): self
    {
        $setFields = $this->setFields;
        $setFields['dbUser'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $this->region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $dbUser,
            workgroupName: $this->workgroupName,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withWorkgroupName(?string $workgroupName): self
    {
        $setFields = $this->setFields;
        $setFields['workgroupName'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $this->region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $workgroupName,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withDbName(?string $dbName): self
    {
        $setFields = $this->setFields;
        $setFields['dbName'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $this->region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $this->workgroupName,
            dbName: $dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withSchemaName(?string $schemaName): self
    {
        $setFields = $this->setFields;
        $setFields['schemaName'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $this->region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $this->workgroupName,
            dbName: $this->dbName,
            schemaName: $schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withTableName(?string $tableName): self
    {
        $setFields = $this->setFields;
        $setFields['tableName'] = true;

        return new self(
            accessKeyId: $this->accessKeyId,
            secretAccessKey: $this->secretAccessKey,
            region: $this->region,
            clusterIdentifier: $this->clusterIdentifier,
            dbUser: $this->dbUser,
            workgroupName: $this->workgroupName,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $tableName,
            setFields: $setFields
        );
    }

    public function jsonSerialize(): mixed
    {
        $data = [
        ];

        if (isset($this->setFields['accessKeyId'])) {
            $data['accessKeyId'] = $this->accessKeyId;
        }
        if (isset($this->setFields['secretAccessKey'])) {
            $data['secretAccessKey'] = $this->secretAccessKey;
        }
        if (isset($this->setFields['region'])) {
            $data['region'] = $this->region;
        }
        if (isset($this->setFields['clusterIdentifier'])) {
            $data['clusterIdentifier'] = $this->clusterIdentifier;
        }
        if (isset($this->setFields['dbUser'])) {
            $data['dbUser'] = $this->dbUser;
        }
        if (isset($this->setFields['workgroupName'])) {
            $data['workgroupName'] = $this->workgroupName;
        }
        if (null !== $this->dbName) {
            $data['dbName'] = $this->dbName;
        }
        if (isset($this->setFields['schemaName'])) {
            $data['schemaName'] = $this->schemaName;
        }
        if (null !== $this->tableName) {
            $data['tableName'] = $this->tableName;
        }

        return \Svix\Utils::newStdClassIfArrayIsEmpty($data);
    }

    /**
     * Create an instance from a mixed obj.
     */
    public static function fromMixed(mixed $data): self
    {
        return new self(
            accessKeyId: \Svix\Utils::deserializeString($data, 'accessKeyId', false, 'RedshiftConfigIn'),
            secretAccessKey: \Svix\Utils::deserializeString($data, 'secretAccessKey', false, 'RedshiftConfigIn'),
            region: \Svix\Utils::deserializeString($data, 'region', false, 'RedshiftConfigIn'),
            clusterIdentifier: \Svix\Utils::deserializeString($data, 'clusterIdentifier', false, 'RedshiftConfigIn'),
            dbUser: \Svix\Utils::deserializeString($data, 'dbUser', false, 'RedshiftConfigIn'),
            workgroupName: \Svix\Utils::deserializeString($data, 'workgroupName', false, 'RedshiftConfigIn'),
            dbName: \Svix\Utils::deserializeString($data, 'dbName', false, 'RedshiftConfigIn'),
            schemaName: \Svix\Utils::deserializeString($data, 'schemaName', false, 'RedshiftConfigIn'),
            tableName: \Svix\Utils::deserializeString($data, 'tableName', false, 'RedshiftConfigIn')
        );
    }

    /**
     * Create an instance from a json string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode(json: $json, associative: true, depth: 512, flags: JSON_THROW_ON_ERROR);

        return self::fromMixed($data);
    }
}
